### ISSUE TITLE
grad_fn() crashed in pytorch 0.3.0

### ISSUE THREAD

**Issue by Giszy** [pytorch/pytorch#4053]
@soumith 
Environment:
unbuntu 16.04+torch-0.3.0.post4-cp35-cp35m-linux_x86_64.whl (from pytorch.org)
windows 10 + pytorch(master) compiled with msvc

```python
import torch
from torch.autograd import Variable
x = Variable(torch.ones(2, 2), requires_grad=True)
y = x + 2
#print(x.grad_fn()) #succeed to catch errors
print(y.grad_fn()) # crashed here!
```
According to internal logic，it should be a failure but not a crash. when I tested the code  in pytorch (0.2.0), it did not crash.

**Comment by albanD** [comment]
Hi,

What are you trying to use `grad_fn` for? You should never be calling it yourself.
Also since x is a leaf variable, `x.grad_fn` is `None`. So the first print should fail.

**Comment by Giszy** [comment]
The code is odd, but it maybe helps to catch errors. To surprise me, it crashed.

**Comment by colesbury** [comment]
The issue is that the generated backwards functions don't do any validation of their input.

I think we should just disable calling grad_fn's manually from Python. Ensuring proper argument parsing everywhere is going to be a pain and I don't see a good use-case for it.

**Comment by albanD** [comment]
@Giszy as the above comment state, this does crash, but these functions should never be called.
If you have a valid use case that requires it that we did not though of, please present it.

@colesbury How can we make these objects not callable from python since they are `THPCppFunction` objects that are also used for the forward conv and all cpp implemented `Function`s?

**Comment by apaszke** [comment]
The old cpp function path will be removed soon, so we can just remove their call methods.

**Comment by kellymarchisio** [comment]
FYI, I got the same segfault running a python interpreter following this [tutorial].(https://hsaghir.github.io/data_science/pytorch_starter/) . The tutorial suggests we call `.creator`, which has been [replaced](https://github.com/pytorch/tutorials/pull/91/files) with `.grad_fn` .  I called `.grad_fn()` out of curiousity, and -> segfault.

**Comment by vishwakftw** [comment]
@apaszke I think this can be closed too.

**Comment by soumith** [comment]
closed via #8774
